File: bindings/csharp/Bison/Attributes.cs
// MIT License © 2025 Binary Dice Games
/**
 * @file Attributes.cs
 * @brief Optional display/documentation metadata for a class, field, or
 *        method -- mirrors `bison_attributes` / `bison.Attributes`.
 */

using System.Runtime.InteropServices;

namespace Bdg.Bison;

/// <summary>Optional display/documentation metadata for a class, field, or method.</summary>
public sealed record Attributes
{
    public string? DisplayName { get; init; }
    public string? Description { get; init; }
    public string? Category { get; init; }
    public bool Obsolete { get; init; }
    public string? ObsoleteMessage { get; init; }
    public bool Required { get; init; }

    /// <summary>Pins the UTF-8-encoded strings for the lifetime of the
    /// returned scope and produces the native struct pointing at them.
    /// Dispose the scope only after the native call that consumed the
    /// pointer has returned.</summary>
    internal NativeAttributesScope ToNative() => new(this);

    internal static Attributes FromNative(in NativeAttributes native) => new()
    {
        DisplayName = MarshalUtf8(native.DisplayName),
        Description = MarshalUtf8(native.Description),
        Category = MarshalUtf8(native.Category),
        Obsolete = native.Obsolete != 0,
        ObsoleteMessage = MarshalUtf8(native.ObsoleteMessage),
        Required = native.Required != 0,
    };

    private static string? MarshalUtf8(nint ptr) => ptr == nint.Zero ? null : Marshal.PtrToStringUTF8(ptr);
}

/// <summary>
/// Owns the UTF-8 buffers backing a <see cref="NativeAttributes"/> struct so
/// they outlive the P/Invoke call that reads them. Mirrors how the Python
/// binding's <c>Attributes._to_c()</c> leans on ctypes' own string-to-bytes
/// keep-alive semantics, made explicit here since C# has no such implicit
/// lifetime extension for manually marshalled pointers.
/// </summary>
internal sealed class NativeAttributesScope : IDisposable
{
    public NativeAttributes Native;
    private readonly List<nint> _owned = new();

    public NativeAttributesScope(Attributes attrs)
    {
        Native = new NativeAttributes
        {
            DisplayName = Alloc(attrs.DisplayName),
            Description = Alloc(attrs.Description),
            Category = Alloc(attrs.Category),
            Obsolete = attrs.Obsolete ? 1 : 0,
            ObsoleteMessage = Alloc(attrs.ObsoleteMessage),
            Required = attrs.Required ? 1 : 0,
        };
    }

    private nint Alloc(string? s)
    {
        if (s is null)
        {
            return nint.Zero;
        }
        var ptr = Marshal.StringToCoTaskMemUTF8(s);
        _owned.Add(ptr);
        return ptr;
    }

    public void Dispose()
    {
        foreach (var ptr in _owned)
        {
            Marshal.FreeCoTaskMem(ptr);
        }
        _owned.Clear();
    }
}

File: bindings/android/bison-lib/src/main/java/com/bdg/bison/Attributes.java
// MIT License © 2025 Binary Dice Games
// Permission is hereby granted, free of charge, to use, copy, modify, and
// distribute this file. See the LICENSE file or
// https://opensource.org/licenses/MIT for details.

package com.bdg.bison;

/**
 * Optional display/documentation metadata for a class, field, or method --
 * mirrors {@code bison_attributes} (see {@code include/bison_c.h}) and the
 * C# binding's {@code Attributes}.
 *
 * <p>Field layout and constructor signature are read directly by
 * {@code bison_jni.cpp} (via cached {@code jfieldID}s and a cached
 * constructor {@code jmethodID}), so both must stay in sync with
 * {@code jni_util.cpp}'s {@code attributes_view}/{@code new_attributes}.
 */
public final class Attributes {
  public final String displayName;
  public final String description;
  public final String category;
  public final boolean obsolete;
  public final String obsoleteMessage;
  public final boolean required;

  public Attributes() {
    this(null, null, null, false, null, false);
  }

  public Attributes(
      String displayName, String description, String category, boolean obsolete, String obsoleteMessage,
      boolean required) {
    this.displayName = displayName;
    this.description = description;
    this.category = category;
    this.obsolete = obsolete;
    this.obsoleteMessage = obsoleteMessage;
    this.required = required;
  }
}
